### DISCUSSION TITLE
[Tycho 3.0.0] Failed creating shared configuration url for null

### DISCUSSION THREAD

**Discussion by birdingman0626** [eclipse-tycho/tycho#1567] [Q&A]
**Hi there,
I'm trying to migrate from Tycho 2.7.5 to 3.0.0 recently and everything works fine with Tycho 2.7.5. What I did to upgrade was just change from 2.7.5 to 3.0.0 of my pom files and settings. I could build binaries with Tycho 3.0.0 but could not run them. The maven gives warnings as below, which will not be in 2.7.5:**

> ...
> Installing org.eclipse.equinox.p2.user.ui.feature.group 2.4.1700.v20220819-1949.
> Installing org.eclipse.ecf.core.ssl.feature.feature.group 1.1.501.v20210409-2301.
> Installing org.eclipse.ecf.filetransfer.ssl.feature.feature.group 1.1.401.v20210409-2301.
> Installing org.eclipse.equinox.p2.extras.feature.feature.group 1.4.1700.v20220819-1949.
> Installing org.eclipse.ecf.filetransfer.httpclient5.feature.feature.group 1.1.600.v20220215-0126.
> Installing org.eclipse.equinox.p2.core.feature.feature.group 1.6.1500.v20220819-1949.
> Installing org.eclipse.ecf.core.feature.feature.group 1.6.1.v20211005-1944.
> Installing org.eclipse.equinox.p2.rcp.feature.feature.group 1.4.1700.v20220819-1949.
> Installing org.eclipse.ecf.filetransfer.feature.feature.group 3.14.1800.v20220215-0126.
> Installing org.eclipse.emf.ecore.feature.group 2.29.0.v20220817-1401.
> Installing org.eclipse.pde.feature.group 3.14.1300.v20220831-1800.
> Installing org.eclipse.wb.core.feature.feature.group 1.10.0.202208311516.
> Installing org.eclipse.wb.core.java.feature.feature.group 1.10.0.202208311516.
> Installing org.eclipse.jdt.feature.group 3.18.1300.v20220831-1800.
> Installing org.eclipse.emf.common.feature.group 2.27.0.v20220817-1401.
> Installing org.eclipse.wb.layout.group.feature.feature.group 1.10.0.202208311516.
> Installing org.eclipse.wb.core.ui.feature.feature.group 1.10.0.202208311516.
> Installing org.eclipse.wb.swt.feature.feature.group 1.10.0.202208311516.
> Installing org.eclipse.wb.rcp.feature.feature.group 1.10.0.202208311516.
> Installing org.eclipse.e4.rcp.feature.group 4.25.0.v20220829-1402.
> [WARNING] [cd799131-2dcb-4c7c-af50-5a729e9f2019][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0] Failed creating shared configuration url for null.
> Operation completed in 11294 ms.
> ...

**and the program throws exceptions when starting:**

> ...
> !SESSION 2022-10-20 17:13:19.576 -----------------------------------------------
> eclipse.buildId=unknown
> java.fullversion=17.0.4.1+1
> JRE 17 Windows 10 amd64-64-Bit Compressed References 20220812_237 (JIT enabled, AOT enabled)
> OpenJ9   - 1d9d16830
> OMR      - b58aa2708
> JCL      - 1f4d354e654 based on jdk-17.0.4.1+1
> BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en_US
> Framework arguments:  -clearPersistedState
> Command-line arguments:  -os win32 -ws win32 -arch x86_64 -clearPersistedState
> 
> !ENTRY org.eclipse.equinox.app 0 0 2022-10-20 17:13:21.364
> !MESSAGE Product com.blackcircle.ui.application.product could not be found.
> 
> !ENTRY org.eclipse.osgi 4 0 2022-10-20 17:13:21.594
> !MESSAGE Application error
> !STACK 1
> java.lang.NullPointerException: Cannot invoke "java.lang.String.length()" because "path" is null
> 	at org.eclipse.emf.common.util.URI$URIPool$PlatformAccessUnit.setValue(URI.java:865)
> 	at org.eclipse.emf.common.util.URI$URIPool.intern(URI.java:1961)
> 	at org.eclipse.emf.common.util.URI.createPlatformPluginURI(URI.java:2750)
> 	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.determineApplicationModelURI(E4Application.java:399)
> 	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.loadApplicationModel(E4Application.java:342)
> 	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.createE4Workbench(E4Application.java:247)
> 	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.start(E4Application.java:155)
> 	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
> 	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
> ...

**Any ideas about it? This is not an urgent issue, as I can stay at 2.x. Many thanks in advance!**

**Comment by laeubi** [comment]
Can you share a reproducer or even provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue?

**Reply by dhendriks** [reply]
I just found out that issue #1242 (MetaRequirementsTest fails on windows) actually has the same message in its description. And, I see the message in this build output: https://github.com/eclipse-tycho/tycho/actions/runs/2816986663/jobs/4448420263:

```
....
Installing tycho-bundles-external 3.0.0.202208080957.
[DEBUG] [2b67472d-bd54-4a49-b950-e2e5cfec77a8][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0-SNAPSHOT] Overall install request is satisfiable
[WARNING] [2b67472d-bd54-4a49-b950-e2e5cfec77a8][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0-SNAPSHOT] Failed creating shared configuration url for null.
Operation completed in 3927 ms.
...
```

It thus looks like I don't have to create a tycho-its for this issue, as it already is visible in existing tests? Is this sufficient for you to reproduce it, and look into what is causing this issue?

(previously accidentally posted here: https://github.com/eclipse-tycho/tycho/issues/1482#issuecomment-1368472287)

**Comment by dhendriks** [comment]
I'm getting these kinds of warnings as well:

```
[INFO] --- tycho-p2-director-plugin:3.0.0:materialize-products (materialize-products) @ org.eclipse.escet.product ---
[INFO] Installing product org.eclipse.escet.product for environment linux/gtk/x86_64 to C:\Programs\Siemens\escet\dev259-1\git\escet\product\org.eclipse.escet.product\target\products\org.eclipse.escet.product\linux\gtk\x86_64\eclipse-escet-dev
Installing org.eclipse.escet.product 0.8.0.dev.
[WARNING] [71251595-7120-4ff3-ad6c-a79222cf10ac][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0] Failed creating shared configuration url for null.
Operation completed in 25444 ms.
[INFO] Installing product org.eclipse.escet.product for environment win32/win32/x86_64 to C:\Programs\Siemens\escet\dev259-1\git\escet\product\org.eclipse.escet.product\target\products\org.eclipse.escet.product\win32\win32\x86_64\eclipse-escet-dev
Installing org.eclipse.escet.product 0.8.0.dev.
[WARNING] [71251595-7120-4ff3-ad6c-a79222cf10ac][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0] Failed creating shared configuration url for null.
Operation completed in 21988 ms.
[INFO] Installing product org.eclipse.escet.product for environment macosx/cocoa/x86_64 to C:\Programs\Siemens\escet\dev259-1\git\escet\product\org.eclipse.escet.product\target\products\org.eclipse.escet.product\macosx\cocoa\x86_64\EclipseESCET-dev.app
Installing org.eclipse.escet.product 0.8.0.dev.
[WARNING] [71251595-7120-4ff3-ad6c-a79222cf10ac][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0] Failed creating shared configuration url for null.
Operation completed in 23624 ms.
```

Edit: I put the wrong story, about different warnings here, by mistake. Edited to be about the warnings that this issue is about.

**Reply by dhendriks** [reply]
Sorry, reported on the wrong warnings here. I've edited my comment. I do get the warnings as discussed in this issue.

**Comment by dhendriks** [comment]
@dev0626 Did you ever find a solution?

**Reply by birdingman0626** [reply]
Hi dhendriks,
I attempted to provide an integrate-test but failed. So I decided to stick with tycho 2.7.5 since I did not find the upgrade attractive in any way. Actually, I'm trying other RCP solutions because I'm having trouble implementing Explorer. Things become tricky as compatibility layers may need to be added and old APIs may need to be used. Anyway, Hope you can figure out the solution.

**Comment by dhendriks** [comment]
> [...] and the program throws exceptions when starting: [...]

I can start the built product without issues. There are no errors in the Error Log.

I think the `Failed creating shared configuration url for null` warning is not related to the `java.lang.NullPointerException: Cannot invoke "java.lang.String.length()" because "path" is null` error you experience.

**Reply by laeubi** [reply]
Yes the NPE should be reported e.g. to the eclipse-platform so they report a usefull error message that allows further investigation instead of running in a runtime exception.

**Comment by dhendriks** [comment]
@laeubi Any idea what could be causing the `[WARNING] [cd799131-2dcb-4c7c-af50-5a729e9f2019][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0] Failed creating shared configuration url for null.` issue?

**Reply by laeubi** [reply]
Previously all error/warnings of the embedded OSGi part where suppressed in Tycho. So it could be that the message was always there but not visible. So if your builds running fine and you can't reproduce it in a minimal example you probably can ignore this.

**Reply by Phillipus** [reply]
I get a few of the `Failed creating shared configuration url for null.` messages for this build: https://github.com/archimatetool/archi

The build itself is fine.

**Reply by laeubi** [reply]
Please make sure to use at least **Tycho 3.0.1**

**Reply by Phillipus** [reply]
Yes, I'm using 3.0.1

**Comment by Bananeweizen** [comment]
https://github.com/eclipse-equinox/p2/issues/219 might be the root cause for all of you.

**Reply by dhendriks** [reply]
https://github.com/eclipse-equinox/p2/issues/219 was fixed in February. I still get these warnings using Tycho 4.0.0-SNAPSHOT. Any idea what is going on? If Tycho not using the new P2 with the fix yet?

**Reply by laeubi** [reply]
As far as I can see it was fixed in equinox and we have upgraded to the latest release but I think one has to find out in what version it was fixed and if Tycho is really using that version to be sure.

**Reply by dhendriks** [reply]
* It is still there in the code, I see: https://github.com/eclipse-equinox/p2/blob/master/bundles/org.eclipse.equinox.frameworkadmin.equinox/src/org/eclipse/equinox/internal/frameworkadmin/equinox/EquinoxFwConfigFileParser.java#L333
* It is not in the one that is part of Eclipse 2023-03, namely 1.2.200.v20220315-2155, see https://www.eclipse.org/downloads/download.php?file=/releases/2023-03/202303151000/plugins/org.eclipse.equinox.frameworkadmin.equinox.source_1.2.200.v20220315-2155.jar
* It is in the one that is part of Eclipse 2023-06, namely 1.2.300.v20230320-0659, see https://www.eclipse.org/downloads/download.php?file=/releases/2023-06/202306141000/plugins/org.eclipse.equinox.frameworkadmin.equinox.source_1.2.300.v20230320-0659.jar

Where can I see whether Tycho is using the one from Eclipse 2023-03 or 2023-06?

**Reply by laeubi** [reply]
Tycho is not using a fixed release but consumes artifact from maven central, so you can look into the dependency:tree of tycho to find out what frameworkadmin is pulled in.

**Reply by dhendriks** [reply]
It seemed we indeed used Eclipse 2023-03, and thus did not yet get the fix. That was when we used Tycho 2.7.5. However, with Eclipse 2023-03 and Tycho 4.0, the problem is also gone now.
